frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of a "bags list": a semi-sorted list where ordering granularity is dictated by
+//! configurable thresholds that delineate the boundaries of bags. It uses a pattern of composite
+//! data structures, where multiple storage items are masked by one outer API. See [`ListNodes`],
+//! [`CounterForListNodes`] and [`ListBags`] for more information.
+//!
+//! The outer API of this module is the [`List`] struct. It wraps all acceptable operations on top
+//! of the aggregate linked list. All operations with the bags list should happen through this
+//! interface.
+
+use crate::Config;
+use codec::{Decode, Encode};
+use frame_election_provider_support::{VoteWeight, VoteWeightProvider};
+use frame_support::{traits::Get, DefaultNoBound};
+use sp_std::{
+	boxed::Box,
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	iter,
+	marker::PhantomData,
+};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+	/// A duplicate id has been detected.
+	Duplicate,
+}
+
+#[cfg(test)]
+mod tests;
+
+/// Given a certain vote weight, to which bag does it belong to?
+///
+/// Bags are identified by their upper threshold; the value returned by this function is guaranteed
+/// to be a member of `T::BagThresholds`.
+///
+/// Note that even if the thresholds list does not have `VoteWeight::MAX` as its final member, this
+/// function behaves as if it does.
+pub(crate) fn notional_bag_for<T: Config>(weight: VoteWeight) -> VoteWeight {
+	let thresholds = T::BagThresholds::get();
+	let idx = thresholds.partition_point(|&threshold| weight > threshold);
+	thresholds.get(idx).copied().unwrap_or(VoteWeight::MAX)
+}
+
+/// Data structure providing efficient mostly-accurate selection of the top N id by `VoteWeight`.
+///
+/// It's implemented as a set of linked lists. Each linked list comprises a bag of ids of
+/// arbitrary and unbounded length, all having a vote weight within a particular constant range.
+/// This structure means that ids can be added and removed in `O(1)` time.
+///
+/// Iteration is accomplished by chaining the iteration of each bag, from greatest to least. While
+/// the users within any particular bag are sorted in an entirely arbitrary order, the overall vote
+/// weight decreases as successive bags are reached. This means that it is valid to truncate
+/// iteration at any desired point; only those ids in the lowest bag can be excluded. This
+/// satisfies both the desire for fairness and the requirement for efficiency.
```

Comment:
I think most of these docs of duplicates and should be `//`, and the only important thing I see missing is a mention to `List` being the ONLY entry point to this module, and the fact that it should be forbidden to use any of the other types.

Reply:
06a05c498c3a26623d9d8dba492725c028de314c && 69ac8c14fadc1476927b6c43a7ad60f94c07e622